pkg/ioutil/container_io.go

```diff
@@ -0,0 +1,175 @@
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package ioutil
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/url"
+	"os"
+	"os/exec"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/containerd/containerd/cio"
+	"github.com/containerd/containerd/defaults"
+	"github.com/containerd/containerd/pkg/process"
+	"github.com/hashicorp/go-multierror"
+)
+
+const binaryIOProcTermTimeout = 12 * time.Second // Give logger process solid 10 seconds for cleanup
+
+// nerdctl io is a basic container IO implementation.
+type ncio struct {
+	cmd     *exec.Cmd
+	config  cio.Config
+	wg      *sync.WaitGroup
+	closers []io.Closer
+	cancel  context.CancelFunc
+}
+
+func (c *ncio) Config() cio.Config {
+	return c.config
+}
+
+func (c *ncio) Wait() {
+	if c.wg != nil {
+		c.wg.Wait()
+	}
+}
+
+func (c *ncio) Close() error {
+
+	var lastErr error
+
+	if c.cmd == nil || c.cmd.Process == nil {
+
+		// Send SIGTERM first, so logger process has a chance to flush and exit properly
+		if err := c.cmd.Process.Signal(syscall.SIGTERM); err != nil {
+			lastErr = fmt.Errorf("failed to send SIGTERM: %w", err)
+
+			if err := c.cmd.Process.Kill(); err != nil {
+				lastErr = multierror.Append(lastErr, fmt.Errorf("failed to kill process after faulty SIGTERM: %w", err))
+			}
+
+		}
+
+		done := make(chan error, 1)
+		go func() {
+			done <- c.cmd.Wait()
+		}()
+
+		select {
+		case err := <-done:
+			return err
+		case <-time.After(binaryIOProcTermTimeout):
+
+			err := c.cmd.Process.Kill()
+			if err != nil {
+				lastErr = fmt.Errorf("failed to kill shim logger process: %w", err)
+			}
+
+		}
+	}
+
+	for _, closer := range c.closers {
+		if closer == nil {
+			continue
+		}
+		if err := closer.Close(); err != nil {
+			lastErr = err
+		}
+	}
+	return lastErr
+}
+
+func (c *ncio) Cancel() {
+	if c.cancel != nil {
+		c.cancel()
+	}
+}
+
+func NewContainerIO(namespace string, logURI string, tty bool, stdin io.Reader, stdout, stderr io.Writer) cio.Creator {
+	return func(id string) (cio.IO, error) {
+
+		// is from https://github.com/containerd/containerd/blob/main/cmd/containerd-shim-runc-v2/process/io.go#L247
+		uri, err := url.Parse(logURI)
+		if err != nil {
+			return nil, err
+		}
+		sr, sw, err := os.Pipe()
+
+		if err != nil {
+			return nil, err
+		}
+		serrr, serrw, err := os.Pipe()
+		if err != nil {
+			return nil, err
+		}
+
+		r, w, err := os.Pipe()
+		if err != nil {
+			return nil, err
+		}
+
+		cmd := process.NewBinaryCmd(uri, id, namespace)
+		cmd.ExtraFiles = append(cmd.ExtraFiles, sr, serrr, w)
+
+		if err := cmd.Start(); err != nil {
```

Comment:
The pipes should be cleaned up on an error

Reply:
Updated with closer to close pipes and stop logger process on errors.